scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/core/service/impl/FileSystemService.java

```diff
@@ -0,0 +1,167 @@
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.context.annotation.PropertySource;
+import org.springframework.stereotype.Service;
+import org.springframework.util.StringUtils;
+
+import java.io.ByteArrayInputStream;
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.file.*;
+import java.util.Objects;
+
+@Service
+@PropertySource("classpath:/config.properties")
+public class FileSystemService implements IFileSystemService {
+
+    @Value("${rootUploadLocation}")
+    private String rootLocationString;
+
+    @Autowired
+    public FileSystemService() {
+    }
+
+    /**
+     * Creates the dataset (and its required directories) on the filesystem.
+     *
+     * @param id       The id of the dataset.
+     * @param username The username of the dataset.
+     * @throws DatasetStorageException The exception thrown if the directory could not be created.
+     */
+    @Override
+    public void addDatasetDirectories(String id, String username) throws DatasetStorageException {
+        // First, the correct directory must be created
+        Path path = createPath(username, id);
+        File directory = path.toFile();
+        // if the Path was created successfully, create the directory
+        try {
+            if (!directory.exists()) {
+                Files.createDirectories(path);
+            }
+        } catch (IOException e) {
+            throw new DatasetStorageException("IOException occurred in addDatasetDirectories", e);
+        } catch (UnsupportedOperationException e) {
+            throw new DatasetStorageException("UnsupportedOperationException occurred in addDatasetDirectories", e);
+        } catch (SecurityException e) {
+            throw new DatasetStorageException("SecurityException occurred in addDatasetDirectories", e);
+        }
+    }
+
+    /**
+     * A helper method to create a Path given a username and dataset name.
+     *
+     * @param username  The username of the owner of the dataset.
+     * @param datasetId The id of the dataset.
+     * @return Returns a Path.
+     */
+    @Override
+    public Path createPath(String username, String datasetId) {
+        Path path;
+        if (datasetId != null) {
+            try {
+                path = Paths.get(this.rootLocationString, username, datasetId);
+                String pathName = StringUtils.cleanPath(Objects.requireNonNull(path).toString());
+                path = Paths.get(pathName);
+                return path;
+            } catch (InvalidPathException e) {
+                throw new DatasetStorageException("An InvalidPathException occurred in createPath with null datasetName", e);
+            }
+        } else {
+            try {
+                path = Paths.get(this.rootLocationString, username);
+                String pathName = StringUtils.cleanPath(Objects.requireNonNull(path).toString());
+                path = Paths.get(pathName);
```

Comment:
the above 2 lines are the same as in the if block and should be consolidated (DRY - don't repeat yourself)

Reply:
the try/catch block could then be around the if/else block which would make it easier to read too